Added API support for local Zonos.

PhialsBasement opened the pull request:
# Add REST API Endpoints

This PR adds FastAPI endpoints to Zonos, allowing programmatic access to the model's functionality alongside the existing Gradio interface.

## Added Features
- `/models` endpoint to list available models
- `/generate` endpoint for text-to-speech generation
- `/speaker_embedding` endpoint for creating speaker embeddings

## Changes
- Added FastAPI integration
- Model responses are streamed as WAV files
- Added Pydantic models for request validation

## Testing
Tested with curl commands:
- GET /models works as expected
- POST /generate successfully generates audio
- POST /speaker_embedding successfully creates embeddings

The implementation reuses existing model management code and runs alongside the Gradio interface on a different port.

PhialsBasement committed:
Update gradio_interface.py

PhialsBasement committed:
Fixed spawn error

PhialsBasement committed:
Added some Stability impovements. Ready for merge when needed.

Steveboy123 commented:
Thank you @PhialsBasement , you are a lifesaver.

zaydek commented:
Just want to mention this thread as relevant for when a teammate comes around to see this PR: https://github.com/Zyphra/Zonos/issues/37.

Ph0rk0z commented:
With OAI endpoint and speakers from folder as returned voices it would work straight away in sillytavern. Unconditional emotions and it would be good "as-is".

PhialsBasement committed:
Create api.py

PhialsBasement committed:
Split API and Gradio

PhialsBasement committed:
Split into API and UI

PhialsBasement commented:
@darkacorn just threw in some of your suggestions, check it out and tell me if its what you were thinking

PhialsBasement committed:
Update api.py

PhialsBasement committed:
Update Dockerfile to accomodate MP3

ther3zz commented:
I'm currently testing the openai endpoint, will report back if I run into any issues!
That being said, it makes sense to include a swagger docs endpoint as well

ther3zz commented:
Has anyone been able to create embeddings? I'm running into this error:

```
{
    "detail": "'int' object has no attribute 'query'"
}
```

PhialsBasement committed:
Update api.py

PhialsBasement commented:
@ther3zz Fixed. Issue was in api.py, i was tryina use .query() on a CUDA stream handle, now its just a normal UNIX timestamp instead.
![image](https://github.com/user-attachments/assets/0aeb00a1-fb0e-4fce-b45e-a90eeb1cbb22)

ther3zz commented:
> @ther3zz Fixed. Issue was in api.py, i was tryina use .query() on a CUDA stream handle, now its just a normal UNIX timestamp instead. ![image](https://private-user-images.githubusercontent.com/155454343/413658866-0aeb00a1-fb0e-4fce-b45e-a90eeb1cbb22.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk3NTMyMTQsIm5iZiI6MTczOTc1MjkxNCwicGF0aCI6Ii8xNTU0NTQzNDMvNDEzNjU4ODY2LTBhZWIwMGExLWZiMGUtNGZjZS1iNDVlLWE5MGVlYjFjYmIyMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxN1QwMDQxNTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MmVjZmI3NjQ2ZDZkNzBmNjZiMTQ3NTJjYzVmM2QxMTY3YmJmZjcwOGU1ZGNjODdjMThmNTZhZmRhNzkzNDlkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.OcHTs60XDBB9hF2SZFEf4qIZBzA49OUaevoC22SlzFo)

Looks like it's working!

ther3zz commented:
Another issue I noticed is that MODEL_CACHE_DIR=/app/models doesnt seem to work. I'm not seeing the models cached there. I see them going here: /root/.cache/huggingface/hub/

PhialsBasement commented:
Whack, ill look into it and see whats going on there

Ph0rk0z commented:
Why can't we just load models from a folder we manually saved? I get that huggingface hub is used for docker, but not all of us are doing that.

Ph0rk0z commented:
I've had to change loading to from_local in gradio and all. The from_pretrained is hijacked away from torch.

mathematicalmichael commented:
hope this helps: HF hub config:
https://huggingface.co/docs/huggingface_hub/package_reference/environment_variables#hfhubcache

PhialsBasement commented:
@ther3zz can you move this to issues tab over on my fork?

Logophoman reviewed (approved):
looks great!

darkacorn commented on api.py:
```diff
@@ -0,0 +1,184 @@
+        codes = model.generate(
+            prefix_conditioning=conditioning,
+            max_new_tokens=86 * 30,
+            cfg_scale=2.0,
```
param or fallback to 2 as default

darkacorn commented on api.py:
```diff
@@ -0,0 +1,184 @@
+            max_new_tokens=86 * 30,
+            cfg_scale=2.0,
+            batch_size=1,
+            sampling_params=dict(min_p=0.15)
```
min_p, top_p, top_k, the 3 unified sampler params, temp , repetition  penality

darkacorn commented on api.py:
```diff
@@ -0,0 +1,184 @@
+            status_code=500,
+            detail=error_msg
+        )
+
```
```
def create_and_save_embedding(file_path: str, voice_name: str):
    wav, sr = torchaudio.load(file_path)
    wav = torch.mean(wav, dim=0, keepdim=True).to(model_manager.device)
    
    transformer_model = model_manager.get_model("transformer")
    with torch.inference_mode():
        speaker_emb = transformer_model.make_speaker_embedding(wav, sr)
        speaker_emb = speaker_emb.to(torch.float16)
    
    embedding_path = audio_manager.voices_dir / f"{voice_name}.pt"
    torch.save(speaker_emb.cpu(), embedding_path)
    logger.info(f"Saved speaker embedding to {embedding_path}")
```

something akin to that could save the tensor - and persist it cross api reboots

ther3zz commented:
> i think there are a few things we should do first before that can be merged - but @Remover88 you can pull that directly even without it beeing merged
> 
> id would like to see the clean audio as prefix (stabilizes the gen ) or to feed in the prefix to stabilze emotions on gen - voice prefix upload overload and stored to disk - the storage maybe as tensor to file so the storage persist cross reboots ( also helps if you scale the api cross nodes - and mount the directly scoped on the userid / s3 )
> 
> additionally there are samplers missing
> 
> just my 2 cents

I 100% agree on persisting files. I accidentally noticed that even though it creates an embedding dir when creating a voice, it doesn't seem to be able to use the voice after a reboot happens (even though the cache dir is mounted to host).

It would also be nice to be able to name the voice and use the custom name when sending generation requests. It looks like right now we have to send the voice ID which is returned when initially creating the voice.

ther3zz commented:
> > > i think there are a few things we should do first before that can be merged - but @Remover88 you can pull that directly even without it beeing merged
> > > id would like to see the clean audio as prefix (stabilizes the gen ) or to feed in the prefix to stabilze emotions on gen - voice prefix upload overload and stored to disk - the storage maybe as tensor to file so the storage persist cross reboots ( also helps if you scale the api cross nodes - and mount the directly scoped on the userid / s3 )
> > > additionally there are samplers missing
> > > just my 2 cents
> > 
> > 
> > I 100% agree on persisting files. I accidentally noticed that even though it creates an embedding dir when creating a voice, it doesn't seem to be able to use the voice after a reboot happens (even though the cache dir is mounted to host).
> > It would also be nice to be able to name the voice and use the custom name when sending generation requests. It looks like right now we have to send the voice ID which is returned when initially creating the voice.
> 
> a fairly easy move .. i just dont want to take op's pr and push that in my self .. i have that stuff already local - nothing major , just semantics and a few quality of life changes - expect for the samplers .. that is kinda crucial

oops, thought your were OP 😁 
thanks for pointing that out!

ProjectBlueBeast commented:
Are there beginner-friendly instructions for the API setup I got the UI to work but I can't get the API part set up.

ther3zz commented:
> Are there beginner-friendly instructions for the API setup I got the UI to work but I can't get the API part set up.

Is the API container running at all or do you mean you're trying to send API requests but that is not working?

ther3zz commented:
@PhialsBasement Any chance on getting these suggestions implemented in your PR?
https://github.com/Zyphra/Zonos/pull/73#issuecomment-2685677927
https://github.com/Zyphra/Zonos/pull/73#issuecomment-2685695625

ProjectBlueBeast commented:
> > Are there beginner-friendly instructions for the API setup I got the UI to work but I can't get the API part set up.
> 
> Is the API container running at all or do you mean you're trying to send API requests but that is not working?

I was trying to hook it up to my sillytavern and got lost

Sturmgewehr444 commented:
The reason this merge should happen as soon as possible is that otherwise we can not implement it into Silltavern. The API is very important.

Sturmgewehr444 commented:
But if we manually clone it Sillytavern would be supporting one specific branch of Zonos that may or may not continue to have its other features or be maintained.

PhialsBasement commented:
> @PhialsBasement Any chance on getting these suggestions implemented in your PR? [#73 (comment)](https://github.com/Zyphra/Zonos/pull/73#issuecomment-2685677927) [#73 (comment)](https://github.com/Zyphra/Zonos/pull/73#issuecomment-2685695625)

ill look into it soon

PhialsBasement commented:
> > Are there beginner-friendly instructions for the API setup I got the UI to work but I can't get the API part set up.
> 
> Is the API container running at all or do you mean you're trying to send API requests but that is not working?

my bad for not including instructions, you need to do 
```
docker build
docker compose up zonos-api
```

after this wait until the api is up and running, it will first download the models and once done it will open the endpoint.

PhialsBasement commented:
FYI, i was on a bit of a break since the last comment i left here. Ill pick up with working on it now so any suggestions should be reiterated in case i miss them.

PhialsBasement committed:
Update README.md

PhialsBasement committed:
Update README.md

PhialsBasement commented:
> But if we manually clone it, Sillytavern would be supporting one specific branch of Zonos that may or may not continue to have its other features or be maintained. We would have to tell everyone : "No, you can't use its latest update, you have to go git switch and then use the API from that particular branch!"

I agree this will become an issue sooner or later if i ever am unable to continue maintaining this.

PhialsBasement committed:
Feat: Add persistent voice storage and custom voice naming

- Implement file-based storage for voice embeddings and metadata
- Add support for custom voice naming during creation
- Enable voice lookup by either name or ID
- Create new /v1/audio/voices endpoint to list saved voices
- Improve reliability with UUID-based voice ID generation
- Enhance error handling with descriptive messages

PhialsBasement commented:
@ther3zz just implemented your suggestions from https://github.com/Zyphra/Zonos/pull/73#issuecomment-2685695625

PhialsBasement committed:
Added Top P, Top K and Min P control for sampling

Ph0rk0z commented:
So one of the concatenation methods has to go into the API. Currently they're targeting the gradio.

PhialsBasement commented:
> So one of the concatenation methods has to go into the API. Currently they're targeting the gradio.

@Ph0rk0z ill look into this when i get back from work tonight

@Sturmgewehr444 im working windows should work just fine but this PR focuses on adding API support primairly. If i have extra time ill look into streamlining it for windows but it will not be a main focus.

Napolitain commented:
why do we have an endpoint for creating a speaker embeddings ?
from my understanding, you make the service stateful. Should we have a stateless service, where we generate the embeddings inside the generation of the audio, and if its repeated, then it will be cached instead? It seems a better design to let the backend handle its resources

ther3zz commented:
> @ther3zz just implemented your suggestions from [#73 (comment)](https://github.com/Zyphra/Zonos/pull/73#issuecomment-2685695625)

Sorry for the delay!

Just tested this and its working perfectly!

YellowRoseCx commented:
would someone mind telling me how I setup the voice IDs with the JSON?

darkacorn commented:
> would someone mind telling me how I setup the voice IDs with the JSON?


voice^^
```
curl -X POST "http://localhost:8000/v1/audio/speech" \
  -H "Content-Type: application/json" \
  -d '{
    "model": "Zyphra/Zonos-v0.1-transformer",
    "input": "Hello, this is a test of the Zonos API.",
    "voice": "voice_12345_0",
    "speed": 1.0,
    "language": "en-us",
    "emotion": {
      "happiness": 1.0
    },
    "response_format": "mp3"
  }' ```

YellowRoseCx commented:
> > would someone mind telling me how I setup the voice IDs with the JSON?
> 
> voice^^
> 
> ```
> curl -X POST "http://localhost:8000/v1/audio/speech" \
>   -H "Content-Type: application/json" \
>   -d '{
>     "model": "Zyphra/Zonos-v0.1-transformer",
>     "input": "Hello, this is a test of the Zonos API.",
>     "voice": "voice_12345_0",
>     "speed": 1.0,
>     "language": "en-us",
>     "emotion": {
>       "happiness": 1.0
>     },
>     "response_format": "mp3"
>   }' ```
> ```

thank you! And I was thinking a good way to improve this overall would be by incorporating 100ms silence that's included in the Zonos/Asset folder as the default value for the "prefix_audio" argument in the speechrequest instead of none because I read somewhere in the file docs or a commit that it increases quality and stability of generations

ther3zz commented:
any luck in getting this merged?
I've been using it and its working really well

kleineluka commented:
Still interested in Zonos having a built-in API like this - any idea if it's possible for merge?

jwoeifjofwefawsfasd commented:
> Still interested in Zonos having a built-in API like this - any idea if it's possible for merge?

+1 here. 

This is easily the best way to garner hype and increase the userbase overnight.

this opens up majority of the usecases where people will actually use a TTS at the current time.

Loving the team's work, looking forward to this.

darkacorn commented:
everyone can pull the fork . the main reason why the core stuff stopped is that v1 is a beta and the team currently focuses on the next version ..

kleineluka commented:
while everyone can pull the fork, there are still conflicts that need dealt with before merging

darkacorn commented:
ya but that should be fairly easy to deal with @PhialsBasement can you alter the pr ?

PhialsBasement commented:
@darkacorn Sure i can remove the gradio interface changes as they are not connected to the api after we switched to using api.py as a dedicated api thingy.

PhialsBasement committed:
Merge branch 'main' into main

PhialsBasement commented:
Should be good now, Let me know if i gotta change anything further.

darkacorn commented:
ill have a chat with gabriel in the afternoon and see if we can get that merged

thank you so far

darkacorn reviewed (approved):
LGTM
